### PR TITLE
Add `WindowBuilder::with_active`

### PR DESCRIPTION
**Pull request by amrbashir**
- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Porting https://github.com/tauri-apps/tao/commit/e42ff07190c37b6b8a7808900133838247a730df

Fixes https://github.com/rust-windowing/winit/issues/2580

### CODE REVIEW

**Review by kchibisov — CHANGES_REQUESTED** [review]
I'm fine with the change, however we should solve the focusing on startup issue first. Given that you don't add any new events, I'd assume that you rely on window being focused.

See this issue
https://github.com/rust-windowing/winit/issues/1077

I'd advise to add a synthetic event(Basically send Focused(false)/Focused(true) where it makes sense) on platform where window is focused/unfocused by default.

By default we should also state that the window is unfocused.

**Comment by amrbashir** [comment]
> I'm fine with the change, however we should solve the focusing on startup issue first. Given that you don't add any new events, I'd assume that you rely on window being focused.

I think on all platforms, a newly spawned window always gains focus, and this method mainly exists to stop that behavior which is useful for a widgets/overlay app where you don't necessarily need focus and merely want to show the window on Top with some content.

> I'd advise to add a synthetic event(Basically send Focused(false)/Focused(true) where it makes sense) on platform where window is focused/unfocused by default.

I am okay with synthesizing a Focus event, should that be implemented in another PR or in this one? 

> By default we should also state that the window is unfocused.

I think the opposite is true, no? a window by default gains focus, unless `with_active` is used with `false`?
or did you mean that users should base their logic on the window not having focus and handle the synthesized `Focus` event instead?

**Comment by amrbashir** [comment]
Looking at #1077, using winit's master branch, I get the correct events order on Windows that matches x11 here https://github.com/rust-windowing/winit/issues/1077#issuecomment-516211759.
Unfortunately, I won't be able to test on macOS or Wayland so I think this should be fixed in another PR, is this considered a hard blocker?

**Comment by kchibisov** [comment]
So what I think we should do is to:

1. Say in docs for `Window::new` that **Focus** state of the window will be delivered via explicit event. And say that initial state of the window is **unfocused**...
2. Emit explicitly whether the window is focused or not from the `new` (this is done on Wayland already).

So if the new window is always focused then you send explicit focus, if it's not focused on start or may not be focused you send unfocused, but be careful here, since we don't want to end up unfocused.

@madsmtm do you know the focus situation on macOS on start? I'd assume it's always focused, no? So we'd need an explicit focus.

**Comment by amrbashir** [comment]
If I understand this correctly the flow should be:
1. create window (unfocused)
2. emit `Focused(false)` event
3. create second window (unfocused)
4. emit `Focused(false)` event
5. emit `Focused(true)` event for the first window

**Comment by kchibisov** [comment]
>emit Focused(true) event for the first window

For the one that is focused by the os. But yes, something like that. If you can guarantee that Os will send `Focused(true)` then it's fine, I guess.

**Comment by amrbashir** [comment]
I tested on Windows and according to this comment https://github.com/rust-windowing/winit/issues/1077#issuecomment-516211759 , that seems to be the current behavior and you said that Wayland synthesize this so only macOS is left which I cannot test.

**Comment by kchibisov** [comment]
@madsmtm ping wrt macOS.

**Comment by madsmtm** [comment]
`windowDidBecomeKey:` is run regardless of how the window became focused, so the `Focused(_)` event is correctly tracked on macOS.

The window is assumed to be unfocused by default, so if we need to match the other platforms here, we should emit a `Focused(false)` event once the window is created (a simple `AppState::queue_event(...)` inside `Window::new` should suffice).

**Comment by madsmtm** [comment]
I've pushed a fix that does this now

**Review by madsmtm — APPROVED** [review]
This looks fine to me.

(Note that, like many things, it won't work on macOS if done outside the event loop, see [the readme](https://github.com/rust-windowing/winit#macos). But that's fine for now).

File: src/window.rs [hunk 1]
```diff
@@ -403,6 +405,17 @@ impl WindowBuilder {
         self
     }
 
+    /// Whether the window will be initially focused or not.
+    ///
+    /// ## Platform-specific:
+    ///
+    /// **Android / iOS / X11 / Wayland:** Unsupported.
```
Comment (kchibisov) [inline src/window.rs:412]:
```suggestion
    /// **Android / iOS / X11 / Wayland / Orbital:** Unsupported.
```

**Comment by kchibisov** [comment]
>I tested on Windows and according to this comment https://github.com/rust-windowing/winit/issues/1077#issuecomment-516211759 for X11 , that seems to be the current behavior and you said that Wayland synthesize this so only macOS is left which I cannot test.

I've retested on X11 and it wasn't like that. Though, I've pushed a fix for that. Be aware that we can't use `CreateNotify` given that for whatever reason Xorg doesn't bother sending it (Or maybe it's an XWayland bug?).

**Review by kchibisov — APPROVED** [review]
I've added X11 impl and slightly changed the docs for the `with_active`. Should be good now.

File: src/platform_impl/windows/window_state.rs [hunk 2]
```diff
@@ -116,6 +116,8 @@ bitflags! {
         /// Drop shadow for undecorated windows.
         const MARKER_UNDECORATED_SHADOW = 1 << 20;
 
+        const MARKER_NO_ACTIVATE = 1 << 18;
```
Comment (msiglreith) [inline src/platform_impl/windows/window_state.rs:119]: Should invert the flag as it's easier to read imo and adjust the logic accordingly - `MARKER_ACTIVATE`

**Pull request merged**